### PR TITLE
feat(SpawnCommand): add spawn command

### PR DESCRIPTION
**Pull request by Mednoob**
Demonstration:

LoggingDemonstration

### CODE REVIEW

**Comment by Hazmi35** [comment]
What's the difference with the regular exec command?

**Comment by Mednoob** [comment]
> What's the difference with the regular exec command?

If you use pm2, there's a command `pm2 logs` right? Using the regular exec command won't give the result since `pm2 logs` will end only if it is `ctrl + c`'d or by closing the terminal. But, with the logs option, it will show the result when the pm2 logs give outputs. Because, the logs option creates new process and then send output to discord whenever the process gets an output.

Here's the demonstration:
PM2Demonstration

**Comment by Hazmi35** [comment]
Is there a way to terminate the logs instance? Showing it's PID and using exec to run `kill <pid> -s 0` is doable, but isn't it better to implement a way to terminate it?

**Comment by Mednoob** [comment]
> Is there a way to terminate the logs instance? Showing it's PID and using exec to run `kill <pid> -s 0` is doable, but isn't it better to implement a way to terminate it?

Yes, and that's what I'm currently thinking about.

Of course, it will use caching system to cache the spawned process. I'm thinking of these ways to end the process:
1. Show its PID and then create an option named `terminate` or something.
2. Like number 1, but with simpler custom ID instead.
3. Create an option named `kill-all` or `terminate-all` to kill **all** the process spawned by the bot instantly.
4. Create some options on the process spawn message using reaction or interaction collector

Or maybe you have any better or simpler idea?

**Comment by Hazmi35** [comment]
> > Is there a way to terminate the logs instance? Showing it's PID and using exec to run `kill <pid> -s 0` is doable, but isn't it better to implement a way to terminate it?
> 
> Yes, and that's what I'm currently thinking about.
> 
> Of course, it will use caching system to cache the spawned process. I'm thinking of these ways to end the process:
> 
> 1. Show its PID and then create an option named `terminate` or something.
> 2. Like number 1, but with simpler custom ID instead.
> 3. Create an option named `kill-all` or `terminate-all` to kill **all** the process spawned by the bot instantly.
> 4. Create some options on the process spawn message using reaction or interaction collector
> 
> Or maybe you have any better or simpler idea?

Come to think of it, shouldn't this be a separate command? SpawnCommand for example
With that we could create something like `$spawn create <name> <shell command>` and `$spawn terminate <name>`
Then, we can store the ChildProcess with the name as the key, and then we can use ChildProcess#kill to terminate it.

**Comment by Hazmi35** [comment]
> > > Is there a way to terminate the logs instance? Showing it's PID and using exec to run `kill <pid> -s 0` is doable, but isn't it better to implement a way to terminate it?
> > 
> > 
> > Yes, and that's what I'm currently thinking about.
> > Of course, it will use caching system to cache the spawned process. I'm thinking of these ways to end the process:
> > 
> > 1. Show its PID and then create an option named `terminate` or something.
> > 2. Like number 1, but with simpler custom ID instead.
> > 3. Create an option named `kill-all` or `terminate-all` to kill **all** the process spawned by the bot instantly.
> > 4. Create some options on the process spawn message using reaction or interaction collector
> > 
> > Or maybe you have any better or simpler idea?
> 
> Come to think of it, shouldn't this be a separate command? SpawnCommand for example
> With that we could create something like `$spawn create <name> <shell command>` and `$spawn terminate <name>`
> Then, we can store the ChildProcess with the name as the key, and then we can use ChildProcess#kill to terminate it.

This way, it wouldn't interfere with ExecCommand, for example. This functionality blocks user who wants to run "logs --name App" from the ExecCommand.

So I think it's best to separate them from ExecCommand

**Comment by Mednoob** [comment]
> > > > Is there a way to terminate the logs instance? Showing it's PID and using exec to run `kill <pid> -s 0` is doable, but isn't it better to implement a way to terminate it?
> > > 
> > > 
> > > Yes, and that's what I'm currently thinking about.
> > > Of course, it will use caching system to cache the spawned process. I'm thinking of these ways to end the process:
> > > 
> > > 1. Show its PID and then create an option named `terminate` or something.
> > > 2. Like number 1, but with simpler custom ID instead.
> > > 3. Create an option named `kill-all` or `terminate-all` to kill **all** the process spawned by the bot instantly.
> > > 4. Create some options on the process spawn message using reaction or interaction collector
> > > 
> > > Or maybe you have any better or simpler idea?
> > 
> > 
> > Come to think of it, shouldn't this be a separate command? SpawnCommand for example
> > With that we could create something like `$spawn create <name> <shell command>` and `$spawn terminate <name>`
> > Then, we can store the ChildProcess with the name as the key, and then we can use ChildProcess#kill to terminate it.
> 
> This way, it wouldn't interfere with ExecCommand, for example. This functionality blocks user who wants to run "logs --name App" from the ExecCommand.
> 
> So I think it's best to separate them from ExecCommand

Ah, you're right. Thank you. Gonna separate them from the ExecCommand.

**Comment by Hazmi35** [comment]
> > > > > Is there a way to terminate the logs instance? Showing it's PID and using exec to run `kill <pid> -s 0` is doable, but isn't it better to implement a way to terminate it?
> > > > 
> > > > 
> > > > Yes, and that's what I'm currently thinking about.
> > > > Of course, it will use caching system to cache the spawned process. I'm thinking of these ways to end the process:
> > > > 
> > > > 1. Show its PID and then create an option named `terminate` or something.
> > > > 2. Like number 1, but with simpler custom ID instead.
> > > > 3. Create an option named `kill-all` or `terminate-all` to kill **all** the process spawned by the bot instantly.
> > > > 4. Create some options on the process spawn message using reaction or interaction collector
> > > > 
> > > > Or maybe you have any better or simpler idea?
> > > 
> > > 
> > > Come to think of it, shouldn't this be a separate command? SpawnCommand for example
> > > With that we could create something like `$spawn create <name> <shell command>` and `$spawn terminate <name>`
> > > Then, we can store the ChildProcess with the name as the key, and then we can use ChildProcess#kill to terminate it.
> > 
> > 
> > This way, it wouldn't interfere with ExecCommand, for example. This functionality blocks user who wants to run "logs --name App" from the ExecCommand.
> > So I think it's best to separate them from ExecCommand
> 
> Ah, you're right. Thank you. Gonna separate them from the ExecCommand.

Note that you don't need to create another PR, just use this one

**Comment by Mednoob** [comment]
@Hazmi35 Done!

File: src/commands/developers/SpawnCommand.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            }
+            if (this.processes.has(name)) {
+                void message.reply("There's a running process with that name. Terminate it first, and then try again.");
+                return;
+            }
+
+            await message.reply(`❯_ ${args.join(" ")}`);
+            const process = spawn(args.shift()!, args, { shell: true, windowsHide: true })
+                .on("spawn", () => {
+                    void message.channel.send("**Log process spawned**");
```
Comment (Hazmi35) [inline src/commands/developers/SpawnCommand.ts:35]:
```suggestion
                    void message.channel.send(`**Process ${name} spawned**`);
```

File: src/commands/developers/SpawnCommand.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                })
+                .on("close", (code, signal) => {
+                    this.processes.delete(name);
+                    void message.channel.send(`**Log process closed with code ${code!}, signal ${signal!}**`);
```
Comment (Hazmi35) [inline src/commands/developers/SpawnCommand.ts:39]:
```suggestion
                    void message.channel.send(`**Process ${name} closed with code ${code!}, signal ${signal!}**`);
```

File: src/commands/developers/SpawnCommand.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                })
+                .on("error", err => {
+                    void message.channel.send(`**An error occured on the log process**\n\`\`\`${err.message}\`\`\``);
```
Comment (Hazmi35) [inline src/commands/developers/SpawnCommand.ts:42]:
```suggestion
                    void message.channel.send(`**An error occured on the process ${name}: **\n\`\`\`${err.message}\`\`\``);
```

**Review by Hazmi35 — CHANGES_REQUESTED** [review]
We should remove the "Log" here because we can use this command other than logging

**Review by Hazmi35 — APPROVED** [review]
Well, you made it. LGTM.

Thanks for your contribution 🎉

**Pull request merged**